subsys/usb/class/cdc_acm.c

```diff
@@ -902,16 +897,34 @@ static int cdc_acm_poll_in(const struct device *dev, unsigned char *c)
 /*
  * @brief Output a character in polled mode.
  *
- * The UART poll method for USB UART is simulated by waiting till
- * we get the next BULK In upcall from the USB device controller or 100 ms.
+ * The poll function looks similar to cdc_acm_fifo_fill() and
+ * tries to do the best to mimic behavior of a hardware UART controller
+ * without flow control.
+ * This function does not block, if the USB subsystem
+ * is not ready, no data is transferred to the buffer, that is, c is dropped.
+ * If the USB subsystem is ready and the buffer is full, the first character
+ * from the tx_ringbuf is removed to make room for the new character.
  */
-static void cdc_acm_poll_out(const struct device *dev,
-				      unsigned char c)
+static void cdc_acm_poll_out(const struct device *dev, unsigned char c)
 {
-	struct cdc_acm_dev_data_t * const dev_data = DEV_DATA(dev);
+	struct cdc_acm_dev_data_t * const dev_data = dev->data;
+
+	if (!dev_data->configured || dev_data->suspended) {
+		LOG_INF("USB device not ready, drop data");
+		return;
+	}
+
+	dev_data->tx_ready = false;
 
-	cdc_acm_fifo_fill(dev, &c, 1);
-	k_sem_take(&dev_data->poll_wait_sem, K_MSEC(100));
+	if (!ring_buf_put(dev_data->tx_ringbuf, &c, 1)) {
+		LOG_INF("Ring buffer full, drain buffer");
+		if (!ring_buf_get(dev_data->tx_ringbuf, NULL, 1)) {
+			LOG_ERR("Failed to drain buffer");
+			return;
+		}
+	}
```

Comment:
Shouldn't there be another call to `ring_buf_put()` done after the room in the buffer is made successfully to actually put the new character there?

Reply:
right, if was first a while, but while could block so I changed it to if but did not think it through, thanks